File: apps/pwa/src/server/api/get_exploration.ts
import { prefixServerOrigin } from '@/global_states/server';
import { request } from '..';

type MusicItem = {
  id: string;
  name: string;
  cover: string;
  coverThumbnail?: string;
  performers: {
    id: string;
    name: string;
  }[];
};
type ArtistItem = {
  id: string;
  name: string;
  photos: { id: string; asset: string; thumbnail?: string; description: string }[];
};
type PublicMusicbillItem = {
  id: string;
  name: string;
  cover: string;
  user: { id: string; nickname: string };
};
type Response = {
  musicList: MusicItem[];
  artistList: ArtistItem[];
  publicMusicbillList: PublicMusicbillItem[];
  recentMusicList: MusicItem[];
  recentArtistList: ArtistItem[];
  recentPublicMusicbillList: PublicMusicbillItem[];
};

const normalizeMusic = (m: MusicItem) => ({
  ...m,
  cover: prefixServerOrigin(m.cover),
  coverThumbnail: prefixServerOrigin(m.coverThumbnail ?? ''),
});
const normalizeArtist = (artist: ArtistItem) => ({
  ...artist,
  photos: artist.photos.map((p) => ({
    ...p,
    asset: prefixServerOrigin(p.asset),
    thumbnail: prefixServerOrigin(p.thumbnail ?? ''),
  })),
});
const normalizeMusicbill = (mb: PublicMusicbillItem) => ({
  ...mb,
  cover: prefixServerOrigin(mb.cover),
});

/**
 * 获取发现内容
 * @author mebtte<i@mebtte.com>
 */
async function getExploration() {
  const data = await request<Response>({
    path: '/api/exploration',
    withToken: true,
  });
  return {
    musicList: data.musicList.map(normalizeMusic),
    artistList: data.artistList.map(normalizeArtist),
    publicMusicbillList: data.publicMusicbillList.map(normalizeMusicbill),
    recentMusicList: data.recentMusicList.map(normalizeMusic),
    recentArtistList: data.recentArtistList.map(normalizeArtist),
    recentPublicMusicbillList:
      data.recentPublicMusicbillList.map(normalizeMusicbill),
  };
}

export default getExploration;

File: apps/flutter/lib/server/api/get_exploration.dart
import '../request.dart';
import '../../models/music.dart';
import '../../utils/prefix_server_origin.dart';

class ExplorationSinger {
  final String id;
  final String name;
  final String? avatar;

  ExplorationSinger({required this.id, required this.name, this.avatar});

  factory ExplorationSinger.fromJson(Map<String, dynamic> json) {
    return ExplorationSinger(
      id: json['id'],
      name: json['name'],
      avatar: prefixServerOrigin(json['avatar']),
    );
  }
}

class ExplorationPublicMusicbill {
  final String id;
  final String name;
  final String? cover;

  ExplorationPublicMusicbill({
    required this.id,
    required this.name,
    this.cover,
  });

  factory ExplorationPublicMusicbill.fromJson(Map<String, dynamic> json) {
    return ExplorationPublicMusicbill(
      id: json['id'],
      name: json['name'],
      cover: prefixServerOrigin(json['cover']),
    );
  }
}

class ExplorationData {
  final List<Music> musicList;
  final List<ExplorationSinger> singerList;
  final List<ExplorationPublicMusicbill> publicMusicbillList;

  ExplorationData({
    required this.musicList,
    required this.singerList,
    required this.publicMusicbillList,
  });

  factory ExplorationData.fromJson(Map<String, dynamic> json) {
    return ExplorationData(
      musicList:
          (json['musicList'] as List<dynamic>?)
              ?.map((m) => Music.fromJson(m))
              .toList() ??
          [],
      singerList:
          (json['singerList'] as List<dynamic>?)
              ?.map((s) => ExplorationSinger.fromJson(s))
              .toList() ??
          [],
      publicMusicbillList:
          (json['publicMusicbillList'] as List<dynamic>?)
              ?.map((mb) => ExplorationPublicMusicbill.fromJson(mb))
              .toList() ??
          [],
    );
  }
}

Future<ExplorationData> getExploration() async {
  final data = await httpGet(path: '/api/exploration', withToken: true);
  return ExplorationData.fromJson(data);
}
